cupy/core/_routines_math.pyx

```diff
@@ -0,0 +1,531 @@
+        __syncthreads();
+
+        for(int i = 1; i <= ${block_size}; i <<= 1){
+            int index = (threadIdx.x + 1) * i * 2 - 1;
+            if (index < (${block_size} << 1)){
+                temp[index] = temp[index] + temp[index - i];
+            }
+            __syncthreads();
+        }
+
+        for(int i = ${block_size} >> 1; i > 0; i >>= 1){
+            int index = (threadIdx.x + 1) * i * 2 - 1;
+            if(index + i < (${block_size} << 1)){
+                temp[index + i] = temp[index + i] + temp[index];
+            }
+            __syncthreads();
+        }
+
+        if(idx0 < n){
+            dst[idx0] = temp[thid];
+        }
+        if(idx1 < n){
+            dst[idx1] = temp[thid + blockDim.x];
+        }
+    }
+    """).substitute(name=name, dtype=dtype, block_size=block_size)
+    module = compile_with_cache(source)
+    return module.get_function(name)
+
+
+@util.memoize(for_each_device=True)
+def _add_scan_blocked_sum_kernel(dtype):
+    name = "add_scan_blocked_sum_kernel"
+    dtype = get_typename(dtype)
+    source = string.Template("""
+    extern "C" __global__ void ${name}(CArray<${dtype}, 1> src_dst){
+        long long n = src_dst.size();
+        unsigned int idxBase = (blockDim.x + 1) * (blockIdx.x + 1);
+        unsigned int idxAdded = idxBase + threadIdx.x;
+        unsigned int idxAdd = idxBase - 1;
+
+        if(idxAdded < n){
+            src_dst[idxAdded] += src_dst[idxAdd];
+        }
+    }
+    """).substitute(name=name, dtype=dtype)
+    module = compile_with_cache(source)
+    return module.get_function(name)
+
+
+cdef ndarray scan(ndarray a, ndarray out=None):
+    """Return the prefix sum(scan) of the elements.
+
+    Args:
+        a (cupy.ndarray): input array.
+        out (cupy.ndarray): Alternative output array in which to place
+         the result. The same size and same type as the input array(a).
+
+    Returns:
+        cupy.ndarray: A new array holding the result is returned.
+
+    """
+    if a._shape.size() != 1:
+        raise TypeError("Input array should be 1D array.")
+
+    cdef Py_ssize_t block_size = 256
+
+    if out is None:
+        out = ndarray(a.shape, dtype=a.dtype)
+    else:
+        if a.size != out.size:
+            raise ValueError("Provided out is the wrong size")
+
+    kern_scan = _inclusive_scan_kernel(a.dtype, block_size)
+    kern_scan(grid=((a.size - 1) // (2 * block_size) + 1,),
+              block=(block_size,),
+              args=(a, out),
+              shared_mem=a.itemsize * block_size * 2)
+
+    if (a.size - 1) // (block_size * 2) > 0:
+        blocked_sum = out[block_size * 2 - 1:None:block_size * 2]
+        scan(blocked_sum, blocked_sum)
+        kern_add = _add_scan_blocked_sum_kernel(out.dtype)
+        kern_add(grid=((a.size - 1) // (2 * block_size),),
+                 block=(2 * block_size - 1,),
+                 args=(out,))
+    return out
+
+
+# Only for test
+def _scan_for_test(a, out=None):
```

Comment:
Added a wrapper function to expose for test purpose, because `scan` function doesn't have to be exposed otherwise.